New API proposal : extension.getContexts() (#334)
hober/web-extensions: proposals/extension_get_contexts.md

# New API: extension.getContexts()

## Background / Summary

Chromium currently has the

> <code>[extension.getViews()](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>

Reviewer: ```suggestion
[`extension.getViews()`](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)
```

API method which allows an extension to get information about the views the
extension is running in. Common reasons for calling this method are: determining

> if a toolbar popup is open, reading and/or modifying a settings page, and etc.

Reviewer: "Etc." is an abbreviation of "et cetera". This phrase comes from Latin where et means "and" and cetera means "the rest". As such, and should not appear before etc. or the expanded meaning would be "and and the rest".

```suggestion
if a toolbar popup is open, reading and/or modifying a settings page, etc.
```

Chromium's implementation of Manifest V2 (MV2) allows for
[background scripts (pages)](https://developer.chrome.com/docs/extensions/mv2/background_pages/)
to call this API. This is possible because these scripts run on our main thread

> which allows them easy access to the Javascript

Reviewer: ```suggestion
which allows them easy access to the JavaScript
```

> [Window objects](http://go/mdn/API/Window#instance_properties) provided by

Reviewer: ```suggestion
[`Window`](http://go/mdn/API/Window#instance_properties) objects provided by
```

> <code>extension.getViews()</code>.

Reviewer: ```suggestion
`extension.getViews()`.
```

## Problem

However, with Chromium’s migration to Manifest V3 (MV3), background pages
[no longer exist](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/);
instead, an extension’s background context is
[service worker](https://developer.chrome.com/docs/workbox/service-worker-overview/)-based.
For technical reasons<sup>[1](#footnotes)</sup>, service workers cannot access

> the [HTMLWindow](http://go/mdn/API/Window#instance_properties) objects that

Reviewer: ```suggestion
the [`HTMLWindow`](http://go/mdn/API/Window#instance_properties) objects that
```

`extension.getViews()` provides and it is not feasible to implement that with
our browser design. Due to this we cannot provide access to the JS context for
these views, but we can allow an extension to query for them and potentially
target them for messaging purposes.

## Solution

> Considering the above situation, we’d like to propose a new (Manifest V3 only)

Reviewer 2: Why is it v3 only? v2 supports service workers in Chrome and Safari.

Author: I'll remove this. It was phrased that way since Chromium doesn't plan to support this for SW in MV2.

method to asynchronously provide metadata about the view that is still useful
for an extension: `extension.getContexts()`. This will allow extension service
workers (or other contexts) to identify the active contexts in the extension.
For example, this can be used to target messages to send using

> <code>[runtime.sendMessage()](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage)</code>,

Reviewer: ```suggestion
[`runtime.sendMessage()`](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage),
```

etc.). Introducing this API will allow an easier migration from MV2.

## API Proposal

This new API would use a similar parameter signature as
`extension.getViews()`but instead accept a new `contextProperties` object.

```
extension.getContexts()(
  contextProperties?: object,
)
```

`contextProperties` will have similar behavior to `extension.getViews()`’s

> <code>[fetchProperties](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>

Reviewer: ```suggestion
[`fetchProperties`](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)
```

object in that it will limit the scope of the query to certain tabs, context
types, and windows. It would be defined as:

```
extension.ContextProperties = {
  tabId? = int,                // Find context by tab id, omitted returns all tabs.
  contextType? = ContextType,  // Find context by type, omitted returns all types.
  windowId? = int,             // Find context by window id, omitted returns all
                               // windows.
}
```

`ContextType` would serve a similar purpose of

> <code>[fetchProperties](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>’s

Reviewer: ```suggestion
[`fetchProperties`](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)'s
```

> [ViewType](https://developer.chrome.com/docs/extensions/reference/extension/#type-ViewType)

Reviewer: ```suggestion
[`ViewType`](https://developer.chrome.com/docs/extensions/reference/extension/#type-ViewType)
```

in order to limit the type of extension contexts returned by the API. It would
be defined as:

```
extension.ContextType = {
  TAB: 'TAB',                               // Tabs the extension is running in.
  POPUP: 'POPUP',                           // Popups created by the extension.

>   SERVICE_WORKER: 'SERVICE_WORKER',         // Service workers for the extension.

Reviewer 2: I would call this `BACKGROUND` to allow Safari and Firefox to support this for background pages in v3. As mentioned above this should also not be limited to v3, and a more generic name is warranted, since v2 supports service workers or pages.

  OFFSCREEN_DOCUMENT: 'OFFSCREEN_DOCUMENT'  // Offscreen documents for the extension.
};
```

It would return an array of new `Context` objects for each frame the extension
is running in:

```
extension.Context = {
  tabId: int,                    // ID of the tab for this view.
  windowId: int,                 // ID of the window this tab is located in.
  documentId: int,               // ID of the DOM document loaded in the tab.
  frameId: int,                  // ID of the frame this context is running for.
  contextType: ContextType,      // Context type -- tab, popup, etc.

>   url: string,                   // The current URL of the tab.

Reviewer 2: Is this the tab URL or the frame's URL? Will is return or non-tab contexts, like popup or background?

Author: Corrected the typo, this would be the frame's URL so I renamed it to `frameUrl`.  In non-tab context like popup (Chromium as an example) popup  would be "chrome-extension://<id>/popup.html".

For a background page or service worker (since there's only one per extension) it doesn't seem helpful to know URL for it (e.g. "chrome-extension://<id>/background.js" so in those cases this would be empty. However, we could include it if there's a need for it? Applying this same logic would mean the field would be empty for content scripts as well, if supported.

>   origin: string,                // The origin of the context. Mostly this is the

Reviewer 2: What do you mean by origin here? Host name? Security origin?

Reviewer: I read this as meaning the concept as a [URL](https://url.spec.whatwg.org/#url-class) instance's [origin attribute](https://url.spec.whatwg.org/#dom-url-origin).

                                 // extension origin.

>   incognito: boolean             // Whether the context is for an incognito

Reviewer 2: Can we get away from browser marketing names in the API? `ephemeral` is a good neutral term for this. Documentation can tie it to the browser's marketing name.

Reviewer: I'm hesitant about using `ephemeral`. While it's technically accurate (private browsing sessions are ephemerial), that label doesn't convey enough about the capability it's referencing. What about `private`? I think that better represents how users think about – and how browser describe – the feature.

Author: `private` works for me, does that work for you too @xeenon?

Reviewer 3: private lgtm.

Reviewer 2: Yes, `private` is good.

Reviewer 4: For new concepts onwards, I advocate for neutral names.
But "incognito" is already widespread in extension APIs.
@xeenon @mukul-p Is the confusion from different names for exactly the same concept worth the introduction of a new name? E.g.

- `extension.inIncognitoContext`
- `tabs.Tab.incognito`
- `windows.Window.incognito`
- `cookie.cookies.incognito`
- `downloads.DownloadItem.incognito`

Secondly, the relevance of `incognito` is that it reflects the extent to which DOM APIs share state across contexts, such as network state (e.g. cookies, cache), to help extensions with respecting the state isolation between different contexts. The "incognito" boolean is not enough for that. In Firefox, `cookieStoreId` is used as a generic way to distinguish contexts with different cookie stores. This name originates from Chrome's cookies extension API, and values are browser-specific opaque strings: Chrome uses "0" and "1", Firefox uses "firefox-default", "firefox-private", "firefox-container-(digits)" (the latter is used for container tabs - https://wiki.mozilla.org/Security/Contextual_Identity_Project/Containers ).

Reviewer 2: That makes sense @Rob--W, I am fine with sticking with `incognito` to keep the API consistent.

Reviewer 3: Agree with Rob that we should be consistent, hence incognito is good.

Compared to _ephemeral_ inPrivate looks better.

                                 // profile.
}
```

`Context` objects will contain an `incognito` value indicating if the Context is
running in an incognito profile. In this area, this brings up the case of an
extension specifying
[split incognito mode](https://developer.chrome.com/docs/extensions/mv3/manifest/incognito/#split).

> In this case we would not return the <code>Context</code>s from the related (but

Reviewer: ```suggestion
In this case we would not return the `Context`s from the related (but
```

different) extension process running for the incognito profile.

> <code>spanning</code> mode would return any <code>Context</code>s running in an

Reviewer: ```suggestion
`spanning` mode would return any `Context`s running in an
```

> incognito profile (with <code>Context.incognito</code> being <code>true</code>)

Reviewer: ```suggestion
incognito profile (with `Context.incognito` being `true`)
```

> in the same shared process. More comments about this are in the Future Work

Reviewer: Link directly to that section.

```suggestion
in the same shared process. More comments about this are in the [Future Work](#future-work)
```

section on this proposal.

## Future Work

### Multi Page Architecture Fields (`DocumentLifeCycle`, `FrameType`, and `parentDocumentId`, and `parentFrameId`)

[Multi Page Architecture](http://doc/1NginQ8k0w3znuwTiJ5qjYmBKgZDekvEPC22q0I4swxQ#heading=h.w1qo2n6sr8wn)

> changed how outermost frames are presented in web navigation in chromium. More

Reviewer: Chromium is a proper noun. The impact of MPArch wasn't limited to the Web Navigation API; it also affected the Tabs API and may have ripple effects on other platform features (e.g. documentId support in the Scripting API).

```suggestion
changed how outermost frames are exposed to extensions in Chromium. More
```

context on how this can impact extensions is
[here](https://developer.chrome.com/blog/extension-instantnav/). Considering
this new architecture it is likely that extensions will want to know what
navigation state (`DocumentLifeCycle`) a `Context` is in, whether the frame’s
`FrameType` is an outermost frame, and other fields like `parentDocumentId`, and
`parentFrameId`. These are fields we would like to include in a future iteration
of this API.

### Asynchronous and Synchronous Version

A potential avenue is that, since this new API is asynchronous whereas
`getViews()` is synchronous, we could explore a synchronous version of this new
API if there’s significant demand for it. As mentioned
previously<sup>[1](#footnotes)</sup>, there are technical limitations that make
it infeasible to synchronously fetch all `Context`s. We could provide two

> methods, one synchronous, and one asynchronous API methods:

Reviewer 2: `await` is good enough for this. Let's not add any more complex sync APIs.

Author: I added a sentence that required the demand to justify why `await` cannot be used. If no others desire this -- I can replace the whole section before commiting mentioning that there is no significant need for this (just for future reference)?

*   `getContexts()`: as defined above and asynchronous. It fetches all
    `Context`s available in the process running for the profile.
*   `getContextsSynchronous()` (name TBD): defined similarly to `getContexts()`,
    but is synchronous. It returns the data for other `Context`s that are
    available synchronously (running in the same process and on the same
    thread). If called from a service worker, it would likely only return the
    service worker `Context`.

> ### Content Script Contexts <a name="content-scripts-fw">Content Script Contexts</a>

Reviewer: GitHub automatically generates an ID for section titles. In this case, the ID will be "content-script-contexts".

```suggestion
### Content Script Contexts
```

[Content scripts](https://developer.chrome.com/docs/extensions/mv3/content_scripts/)
run in a separate
[Renderer](https://developer.chrome.com/blog/inside-browser-part3/) (and

> process) separate from the extension process. This proposal would not return

Reviewer: "a separate Renderer separate from" is redundant.

```suggestion
process) from the extension process. This proposal would not return
```

those `Context`s since this adds significant complexity to the design in
querying all Renderer processes in the browser to find content scripts. However,
extensions could find it useful to have this `Context` metadata for places where

> the extension is executing javascript in a web page. It could be justified in

Reviewer: ```suggestion
the extension is executing JavaScript in a web page. It could be justified in
```

the future if extension developers would have a significant need.

### Messaging APIs Support `Context`s as Targets

This new API now provides the ability to consider specifying targets via
`Context`s to messaging APIs like `runtime.sendMessage()` or
`tabs.sendMessage()`. With this support it would avoid
[various workarounds](https://github.com/w3c/webextensions/pull/334#issuecomment-1343255899)
needed to target specific pages with messages. `tabs.sendMessage()` accepting
`Context`s would require work in retrieving content script `Context`s explained

> [above](#content-scripts-fw).

Reviewer: Update to match the auto-generated ID for this section's title.

```suggestion
[above](#content-script-contexts).
```

> ## <a name="footnotes">Footnotes</a>

Reviewer: GitHub will automatically generate a `footnotes` anchor for this header.

```suggestion
## Footnotes
```

<sup>1</sup>: Non-main threads in a
[Renderer](https://developer.chrome.com/blog/inside-browser-part3/) (where
service workers run in Chromium) cannot access DOM concepts directly (they are
only accessible from the main browser thread). Service workers run in these

> threads therefore they do not have access to the Javascript

Reviewer: ```suggestion
threads therefore they do not have access to the JavaScript
```

potentially target them for messaging purposes.

## Solution

Considering the above situation, we’d like to propose a new `extension` API
method to asynchronously provide metadata about the view that is still useful
for an extension: `extension.getContexts()`. This will allow extension background scripts to identify the active contexts in the extension.
For example, this can be used to target messages to send using
[`runtime.sendMessage()`](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage),
etc.). Introducing this API will allow an easier migration from MV2.

## API Proposal

This new API would use a similar parameter signature as
`extension.getViews()`but instead accept a new `contextProperties` object.

```
extension.getContexts()(
  contextProperties?: object,
)
```

`contextProperties` will have similar behavior to `extension.getViews()`’s
[`fetchProperties`](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)
object in that it will limit the scope of the query to certain tabs, context
types, and windows. It would be defined as:

```
extension.ContextProperties = {

>   tabId? = int,                // Find context by tab id, omitted returns all tabs.

Reviewer 4: I suggest to rephrase "returns all tabs" - there are contexts that are not tabs. Similarly for windows".

API if there’s significant demand for it. That demand would need to justify why
the usage of `await` on the existing asynchronous API isn't sufficient. As
mentioned previously<sup>[1](#footnotes)</sup>, there are technical limitations
that make it infeasible to synchronously fetch all `Context`s. We could provide
two methods, one synchronous, and one asynchronous API methods:

*   `getContexts()`: as defined above and asynchronous. It fetches all
    `Context`s available in the process running for the profile.
*   `getContextsSynchronous()` (name TBD): defined similarly to `getContexts()`,
    but is synchronous. It returns the data for other `Context`s that are
    available synchronously (running in the same process and on the same
    thread). If called from a background script, it would likely only return the
    background script `Context`.

### Content Script Contexts

[Content scripts](https://developer.chrome.com/docs/extensions/mv3/content_scripts/)
run in a separate
[Renderer](https://developer.chrome.com/blog/inside-browser-part3/) (and
process) from the extension process. This proposal would not return those

> `Context`s since this adds significant complexity to the design in querying all

Reviewer 4: s/design/implementation/.

Renderer processes in the browser to find content scripts. However, extensions
could find it useful to have this `Context` metadata for places where the
extension is executing JavaScript in a web page. It could be justified in the
future if extension developers would have a significant need. If implemented the
`frameUrl` field would be empty as providing a URL to the content script (within
the extension) doesn't have a use case.

### Messaging APIs Support `Context`s as Targets

This new API now provides the ability to consider specifying targets via
`Context`s to messaging APIs like `runtime.sendMessage()` or
`tabs.sendMessage()`. With this support it would avoid
[various workarounds](https://github.com/w3c/webextensions/pull/334#issuecomment-1343255899)
needed to target specific pages with messages. `tabs.sendMessage()` accepting
`Context`s would require work in retrieving content script `Context`s explained
[above](#content-script-contexts).

## Open Questions

### Include `sandbox` Pages and Iframes as `Context`s?

`sandbox` documents and iframes have an
[opaque origin](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin-opaque),

> should we return them? What if they come from within the extension (or outside

Reviewer 4: IMO: no. Sandboxed documents have a `"null"` origin and don't have access to extension APIs. They're hardly different from regular sandboxed iframes from the web.
